### PR TITLE
Add ability to set "essential" field on containers

### LINKED ISSUE #267: Support ECS specific fields that are not supported in compose
`ecs-cli compose` does not support ECS specific parameters that are in compose including essential, Task role ARN, Network mode, etc.

### PR DESCRIPTION
**Pull request by SoManyHs**
Resolves #267.

### CODE REVIEW

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 1]
```diff
@@ -85,6 +92,24 @@ func ConvertToTaskDefinition(taskDefinitionName string, context *project.Context
...
 
+	ecsParams, err := readECSParams(ecsParamsFileName)
+	if err != nil {
+		return nil, err
+	}
+
+	var networkMode string
+	var ecsContainerDefs ContainerDefs
+
+	if ecsParams != nil {
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: I should have asked about this in the last PR, but why is `readECSParams()` allowed to return nil, nil? Your code implies that ecsParams could be nil, and the error could be nil, and this is not an error condition. Why?

**Comment by SoManyHs** [comment]
With following docker-compose file:

```
~/sandbox/test_fields$ cat docker-compose.yml
version: '2'
services:
  wordpress:
    image: wordpress
    cpu_shares: 100
    mem_limit: 52428800
    ports:
      - "80:80"
  mysql:
    image: mysql
    cpu_shares: 100
    mem_limit: 524288000
    environment:
      MYSQL_ROOT_PASSWORD: foobar
```

Default behavior maintained:
```
~/sandbox/test_fields$ cat !$
cat ecs-params.yml
version: 1
task_definition:
  ecs_network_mode: host


~/sandbox/test_fields$ ecs-cli compose up --cluster test-ecs-fields
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0000] Using ECS task definition                     TaskDefinition="ecscompose-test_fields:25"
INFO[0000] Starting container...                         container="9dd66138-952c-4cbd-b413-329fefeef485/mysql"
INFO[0000] Starting container...                         container="9dd66138-952c-4cbd-b413-329fefeef485/wordpress"
INFO[0000] Describe ECS container status                 container="9dd66138-952c-4cbd-b413-329fefeef485/mysql" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:25"
INFO[0000] Describe ECS container status                 container="9dd66138-952c-4cbd-b413-329fefeef485/wordpress" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:25"
INFO[0006] Started container...                          container="9dd66138-952c-4cbd-b413-329fefeef485/mysql" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:25"
INFO[0006] Started container...                          container="9dd66138-952c-4cbd-b413-329fefeef485/wordpress" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:25"

~/sandbox/test_fields$ aws ecs describe-task-definition --task-definition ecscompose-test_fields:25
{
    "taskDefinition": {
        "status": "ACTIVE",
        "networkMode": "host",
        "family": "ecscompose-test_fields",
        "placementConstraints": [],
        "requiresAttributes": [
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.17"
            },
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.18"
            }
        ],
        "volumes": [],
        "taskDefinitionArn": "arn:aws:ecs:us-west-2:<aws_account>:task-definition/ecscompose-test_fields:25",
        "containerDefinitions": [
            {
                "dnsSearchDomains": [],
                "environment": [
                    {
                        "name": "MYSQL_ROOT_PASSWORD",
                        "value": "foobar"
                    }
                ],
                "readonlyRootFilesystem": false,
                "name": "mysql",
                "links": [],
                "mountPoints": [],
                "image": "mysql",
                "privileged": false,
                "essential": true,
                "portMappings": [],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [],
                "ulimits": [],
                "memory": 500,
                "command": [],
                "extraHosts": [],
                "cpu": 100,
                "volumesFrom": [],
                "dockerLabels": {}
            },
            {
                "dnsSearchDomains": [],
                "environment": [],
                "readonlyRootFilesystem": false,
                "name": "wordpress",
                "links": [],
                "mountPoints": [],
                "image": "wordpress",
                "privileged": false,
                "essential": true,
                "portMappings": [
                    {
                        "protocol": "tcp",
                        "containerPort": 80,
                        "hostPort": 80
                    }
                ],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [],
                "ulimits": [],
                "memory": 50,
                "command": [],
                "extraHosts": [],
                "cpu": 100,
                "volumesFrom": [],
                "dockerLabels": {}
            }
        ],
        "revision": 25
    }
}

```
With essential specified:
```
~/sandbox/test_fields$ cat ecs-params.yml
cat ecs-params.yml
version: 1
task_definition:
  ecs_network_mode: host
  services:
    mysql:
      essential: true

~/sandbox/test_fields$ ecs-cli compose up --cluster test-ecs-fields
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
INFO[0000] Using ECS task definition                     TaskDefinition="ecscompose-test_fields:23"
INFO[0000] Starting container...                         container="7d5c43c0-bc62-4788-af4a-041e0823e58a/mysql"
INFO[0000] Starting container...                         container="7d5c43c0-bc62-4788-af4a-041e0823e58a/wordpress"
INFO[0000] Describe ECS container status                 container="7d5c43c0-bc62-4788-af4a-041e0823e58a/wordpress" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:23"
INFO[0000] Describe ECS container status                 container="7d5c43c0-bc62-4788-af4a-041e0823e58a/mysql" desiredStatus=RUNNING lastStatus=PENDING taskDefinition="ecscompose-test_fields:23"
INFO[0006] Started container...                          container="7d5c43c0-bc62-4788-af4a-041e0823e58a/wordpress" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:23"
INFO[0006] Started container...                          container="7d5c43c0-bc62-4788-af4a-041e0823e58a/mysql" desiredStatus=RUNNING lastStatus=RUNNING taskDefinition="ecscompose-test_fields:23"

~/sandbox/test_fields$ aws ecs describe-task-definition --task-definition ecscompose-test_fields:23
{
    "taskDefinition": {
        "status": "ACTIVE",
        "networkMode": "host",
        "family": "ecscompose-test_fields",
        "placementConstraints": [],
        "requiresAttributes": [
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.17"
            },
            {
                "name": "com.amazonaws.ecs.capability.docker-remote-api.1.18"
            }
        ],
        "volumes": [],
        "taskDefinitionArn": "arn:aws:ecs:us-west-2:<aws_account>:task-definition/ecscompose-test_fields:23",
        "containerDefinitions": [
            {
                "dnsSearchDomains": [],
                "environment": [
                    {
                        "name": "MYSQL_ROOT_PASSWORD",
                        "value": "foobar"
                    }
                ],
                "readonlyRootFilesystem": false,
                "name": "mysql",
                "links": [],
                "mountPoints": [],
                "image": "mysql",
                "privileged": false,
                "essential": true,
                "portMappings": [],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [],
                "ulimits": [],
                "memory": 500,
                "command": [],
                "extraHosts": [],
                "cpu": 100,
                "volumesFrom": [],
                "dockerLabels": {}
            },
            {
                "dnsSearchDomains": [],
                "environment": [],
                "readonlyRootFilesystem": false,
                "name": "wordpress",
                "links": [],
                "mountPoints": [],
                "image": "wordpress",
                "privileged": false,
                "essential": false,
                "portMappings": [
                    {
                        "protocol": "tcp",
                        "containerPort": 80,
                        "hostPort": 80
                    }
                ],
                "dnsServers": [],
                "dockerSecurityOptions": [],
                "entryPoint": [],
                "ulimits": [],
                "memory": 50,
                "command": [],
                "extraHosts": [],
                "cpu": 100,
                "volumesFrom": [],
                "dockerLabels": {}
            }
        ],
        "revision": 23
    }
}
```

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 1 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: Because you might not have an `ecs-params.yml` file, in which case reading it shouldn't return anything and shouldn't be an error.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 2]
```diff
@@ -602,3 +629,11 @@ func SortedGoString(v interface{}) (string, error) {
 	}
 	return string(b), nil
 }
+
+func areContainersAllEssential(ecsContainerDefs ContainerDefs) bool {
+	// NOTE: This is a silly hack for now, since the only field in a
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: Can you either change this comment or add a new one that explains why this function is necessary. I found this comment a bit confusing.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 3]
```diff
@@ -602,3 +629,11 @@ func SortedGoString(v interface{}) (string, error) {
...
+	// ContainerDef is essential. If/when there are other fields, we would
+	// want to change the logic for determining whether the essential field
+	// in all Containers was specified.
+	return len(ecsContainerDefs) == 0
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: Looking at this made me wonder if there is an edge case that we haven't considered. What if you have an ECS Params file like this:

```
task_definition:
  task_role_arn: string
  ecs_network_mode: string # valid values: "none", "host", "bridge", and custom ECS network modes.
  services:
    mysql:
    wordpress:
    redis:
```

This should probably be an error, but I have a suspicion the YAML library my parse it and not throw an error- and that this will cause odd behavior in your code.

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 1 above)
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: Ah I see. This makes me think though that may be the processing of the ECS Params file should be moved to its own function. I think that might make the code easier to read. This is just a suggestion though.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 4]
```diff
@@ -99,24 +124,20 @@ func ConvertToTaskDefinition(taskDefinitionName string, context *project.Context
 		containerDef := &ecs.ContainerDefinition{
 			Name: aws.String(name),
 		}
-		if err := convertToContainerDef(context, serviceConfig, volumes, containerDef); err != nil {
-			return nil, err
+
+		// Check if essential is specified for the container
+		ecsContainerDef := &ContainerDef{}
+		if cd, ok := ecsContainerDefs[name]; ok {
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: Fun Golang thing that I think will let you reduce this all to 1 line. If you only capture the first return from accessing a map, and the key does not exist in the map, then you get the default value of the value type. This means that:

`ecsContainerDef := ecsContainerDefs[name]`

Should give you an empty struct if name is not a valid key.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 5]
```diff
@@ -99,24 +124,20 @@ func ConvertToTaskDefinition(taskDefinitionName string, context *project.Context
...
+			ecsContainerDef = &cd
 		}
-		containerDefinitions = append(containerDefinitions, containerDef)
-	}
 
-	ecsParams, err := readECSParams(ecsParamsFileName)
-	if err != nil {
-		return nil, err
-	}
+		allEssential := areContainersAllEssential(ecsContainerDefs)
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: So in the old CLI, we didn't have essential, and so field was auto-set to true by control plane. 

As I understand your code, if the person doesn't specify essential for any containers then all are still set to true.

BUT, what if I set it to be false for only one container? Ie, what if I have 10 service definitions in my docker-compose file, but I have an ECS Params file like this:
```
services:
     mysql:
          essential false
```

And my docker compose file is like this:
```
version: "2"
services:
  web:
    build: web
    command: python app.py
    ports:
     - "5000:5000"
    volumes:
     - ./web:/code
    links:
     - redis
    environment:
  redis:
    image: redis
  # agent section
  mysql:
    image: mysql
    links:
     - web
```

If I understand your code correctly, this would end up setting all of the containers to have essential be false. This is potentially fine- I don't think its a bug. But in a way you have changed the default for essential from True to False- I think we should talk about that in person.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 6]
```diff
@@ -99,24 +124,20 @@ func ConvertToTaskDefinition(taskDefinitionName string, context *project.Context
 		containerDef := &ecs.ContainerDefinition{
 			Name: aws.String(name),
 		}
-		if err := convertToContainerDef(context, serviceConfig, volumes, containerDef); err != nil {
-			return nil, err
+
+		// Check if essential is specified for the container
+		ecsContainerDef := &ContainerDef{}
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: See my other comment, doing this kind of makes False be the default for the essential- and I suspect that the default should be True. Let's discuss this in person.

**Comment by jmahowald** [comment]
Why not have additional fields like essential be done through labels.  That way it's guaranteed  to be compatible with any other consumers of a docker compose ?
```yaml

labels:
     aws-ecs.essential: true
```

**Comment by jmahowald** [comment]
Why not use a label to maintain compatibility generally, e.g.
```yaml
labels:
   aws-ecs.essential: true
```

**Comment by jmahowald** [comment]
Why not use labels to leave compatibility with other consumers of compose as an annotation. e.g.
```yaml
labels:
   aws-ecs.essential: true
```

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 1 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: The processing of the file is already in its own function. It's called `readECSParams`.

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 4 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: Hm, interesting suggestion. I think I prefer this idiom of checking whether an element is in a map though -- seems more clear to me!

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 5 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: That's a great point -- I hadn't considered that edge case. It's true that this would set all containers to be false, and you would see an error like this:

```
~/sandbox/test_fields$ ecs-cli compose up --cluster test-ecs-fields
WARN[0000] Skipping unsupported YAML option...           option name=networks
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=mysql
WARN[0000] Skipping unsupported YAML option for service...  option name=networks service name=wordpress
ERRO[0000] Error registering task definition             error="ClientException: Task definition doesn't have any essential container.\n\tstatus code: 400, request id: cc373549-b46b-11e7-8f7f-e3294f673e50" family="ecscompose-test_fields"
ERRO[0000] Create task definition failed                 error="ClientException: Task definition doesn't have any essential container.\n\tstatus code: 400, request id: cc373549-b46b-11e7-8f7f-e3294f673e50"
FATA[0000] ClientException: Task definition doesn't have any essential container.
	status code: 400, request id: cc373549-b46b-11e7-8f7f-e3294f673e50
```

So you're right in that it's not a bug per se, but it's not a great user experience and is a good argument for making the `areContainersAllEssential` function less trivial =P

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 6 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: The default should not be true because that would sort of defeat the purpose of the feature. I suppose it comes down to whether we are assuming all containers are not essential unless otherwise specified, or that they are essential unless otherwise specified. I was going with the former, but on closer reading of #267 it looks like customers tend to want to mark one-off containers as non-essential. I'll change the code accordingly.

**Comment by SoManyHs** [comment]
Hi @jmahowald, 

Thanks for your suggestion. The design for this feature was discussed [here](https://github.com/aws/amazon-ecs-cli/blob/master/proposals/support_custom_ecs_fields.md).  While we are definitely prioritizing maintaining our compose APIs for our existing customers whose workflows center around docker-compose, we also wanted to keep fields specific to an ECS Task Definition separate from a the docker-compose file. We felt this would clearly communicate which fields are specific to ECS and not docker-compose, so that those using docker-compose files would not have to worry about parsing errors on fields docker-compose doesn't support but are valid in a Task Definition. It also allows for easier extensibility on Task Definition fields in the future.

Hope that helps. Let me know if you have further questions!

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 3 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: This will actually set "essential" to false for all those services. Tests now reflect that. Not sure if that is desirable behavior, but it doesn't throw a YAML parsing error.

Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: Okay, I think that's acceptable. I mostly just wanted to make sure this wouldn't cause your code to get a null pointer exception or something like that.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 7]
```diff
@@ -630,10 +637,30 @@ func SortedGoString(v interface{}) (string, error) {
...
+	nonEssentialCount := 0
+
+	for _, containerDef := range ecsParamsContainerDefs {
+		if !containerDef.Essential {
+			nonEssentialCount += 1
+		}
+	}
+
+	// 'count' is the total number of containers specified in the service config
+	if nonEssentialCount == count {
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: If you want, you could change these ending lines to 

`return nonEssentialCount != count`

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 8]
```diff
@@ -344,11 +349,13 @@ func convertToContainerDef(context *project.Context, inputCfg *config.ServiceCon
 		outputContDef.LinuxParameters.Capabilities.SetDrop(aws.StringSlice(inputCfg.CapDrop))
 	}
 
-	if allEssential == true {
-		outputContDef.SetEssential(true)
-	} else if ecsContainerDef != nil {
+	if ecsContainerDef != nil {
+	// fmt.Printf("CONVERTING CONTAINER_DEF: %+v, TO ECS CONTAINER_DEF: %+v\n", outputContDef, ecsContainerDef)
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: ha ha you should probably remove this commented out code :P

File: ecs-cli/modules/utils/compose/convert_task_definition_test.go [hunk 9]
```diff
@@ -136,14 +142,221 @@ task_definition:
...
+	ecsParamsFileName := tmpfile.Name()
+
+	taskDefinition, err := convertToTaskDefWithEcsParamsInTest(t, []string{"mysql", "wordpress"}, &config.ServiceConfig{}, "", ecsParamsFileName)
+
+	containerDefs := taskDefinition.ContainerDefinitions
+	mysql := findContainerByName("mysql", containerDefs)
+	wordpress := findContainerByName("wordpress", containerDefs)
+
+	if assert.NoError(t, err) {
+		assert.False(t, aws.BoolValue(mysql.Essential), "Expected essential value to match")
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition_test.go]: I'd suggest changing all of the assert messages to be more descriptive, like in this case "Expected essential to be False" or may be "Expected essential on mysql container to be True." If someone ever changes this code and screws up the logic, seeing how the test cases are failing (true vs false being set incorrectly) will make it faster to debug.

File: ecs-cli/modules/utils/compose/convert_task_definition.go [hunk 10]
```diff
@@ -323,6 +349,14 @@ func convertToContainerDef(context *project.Context, inputCfg *config.ServiceCon
 		outputContDef.LinuxParameters.Capabilities.SetDrop(aws.StringSlice(inputCfg.CapDrop))
 	}
 
+	if ecsContainerDef != nil {
+	// fmt.Printf("CONVERTING CONTAINER_DEF: %+v, TO ECS CONTAINER_DEF: %+v\n", outputContDef, ecsContainerDef)
```
Comment (PettitWesley) [inline ecs-cli/modules/utils/compose/convert_task_definition.go]: ha ha you should probably remove this commented out code here and below :P

Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: oops. 😳

File: ecs-cli/modules/utils/compose/convert_task_definition.go (hunk 7 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/convert_task_definition.go]: oh duh. good catch :)
